sync-core/src/main/java/com/cloudant/sync/query/Index.java

```
@@ -0,0 +1,165 @@
+
+    private static final String DEFAULT_TOKENIZER = "simple";
+
+    private static final List<String> validSettings = Arrays.asList(TOKENIZE);
+
+    protected final List<Object> fieldNames;
+
+    protected final String indexName;
+
+    protected final String indexType;
+
+    protected final Map<String, String> indexSettings;
+
+    private Index(List<Object> fieldNames,
+                  String indexName,
+                  String indexType,
+                  Map<String, String> indexSettings) {
+        this.fieldNames = fieldNames;
+        this.indexName = indexName;
+        this.indexType = indexType;
+        this.indexSettings = indexSettings;
+    }
+
+    public static Index setUp(List<Object> fieldNames, String indexName) {
+        // JSON index type is the default
+        return setUp(fieldNames, indexName, JSON_TYPE);
+    }
+
+    public static Index setUp(List<Object> fieldNames, String indexName, String indexType) {
+        return setUp(fieldNames, indexName, indexType, null);
+    }
+
+    /**
+     * The setUp method handles index specific validation and ensures that the constructed
+     * Index object is valid.
+     *
+     * @param fieldNames the field names in the index
+     * @param indexName the index name
+     * @param indexType the index type (json or text)
+     * @param indexSettings the optional settings used to configure the index.
+     *                      Only supported parameter is 'tokenize' for text indexes only.
+     * @return the Index object or null if arguments passed in were invalid.
+     */
+    public static Index setUp(List<Object> fieldNames,
+                              String indexName,
+                              String indexType,
+                              Map<String, String> indexSettings) {
+        if (fieldNames == null || fieldNames.isEmpty()) {
+            logger.log(Level.SEVERE, "No field names were provided.");
+            return null;
```

I don't think we should be logging an error when the user passes us information which invalid. I would also consider this being a programming error, I would throw an InvalidArgumentException here and for the similar expressions below